As a user, I would like to be able to generate a forecast progression plot

---


Author Name: **Hank** (Hank)
Original Redmine Issue: 45682, https://vlab.noaa.gov/redmine/issues/45682
Original Date: 2018-01-19


---

A forecast progression plot merely displays forecast time series by T0 (i.e., T0 in the legend) in a time series plot with observations displayed.  Alex has some examples for the IOEP, which I'll attach later when we start to work on this. 

Most forecast systems, including CHPS/FEWS, can generate forecast progression plots.  Such plots, when also displaying observations, will likely show all of the observations and forecasts, not just those limited to pairs, and at their original time scale, though they may include features to adjust that scale.  In other words, each time series will be loaded independently of the others and then displayed.

For the WRES forecast progression plot, based on an email exchange with James, I propose that we include only paired data in an evaluation version of the forecast progression plot.  Thus, the plot would reflect the time series of pairs handed off to WRES for the pooled windows and in the desired time scale of the WRES output (i.e., after aggregating).  

Obviously, the exact design for the plot is up for debate and I'm certainly open to other ideas.

Scheduled this for SR1 and prioritiy normal, reflecting that this may be slipped if we don't have time to do it.  I've added Alex and James as watchers for now.

Hank







---


Redmine related issue(s): [118363](https://vlab.noaa.gov/redmine/issues/118363)

---

Running as a stand alone, the ABRN1 streamflow evaluation succeeded in 4m 36s, generating a scatter plot, time series plot, and spaghetti plot. All three look good.

But when I add `svg` as an output, I get this exception:

```
2025-08-08T11:50:01.676+0000 107078 [main] ERROR Main - Operation 'execute' completed unsuccessfully
wres.pipeline.InternalWresException: Could not complete project execution
        at wres.pipeline.Evaluator.evaluate(Evaluator.java:409)
        at wres.pipeline.Evaluator.evaluate(Evaluator.java:181)
        at wres.Functions.evaluate(Functions.java:135)
        at wres.Functions.evaluate(Functions.java:97)
        at wres.helpers.MainUtilities.call(MainUtilities.java:107)
        at wres.Main.completeExecution(Main.java:206)
        at wres.Main.main(Main.java:126)
Caused by: wres.pipeline.WresProcessingException: Encountered an error while processing evaluation 'fQFPkJxjAlIB6g6boQ5s_AUjf9U': 
        at wres.pipeline.Evaluator.evaluate(Evaluator.java:856)
        at wres.pipeline.Evaluator.evaluate(Evaluator.java:380)
        ... 6 common frames omitted
Caused by: wres.events.EvaluationFailedToCompleteException: Failed to complete evaluation fQFPkJxjAlIB6g6boQ5s_AUjf9U due to a subscriber error. The failing subscribers are [-kOOBr7whZvwa65_oJHL_gA-GME].
        at wres.events.EvaluationMessager.await(EvaluationMessager.java:788)
        at wres.pipeline.Evaluator.evaluate(Evaluator.java:802)
        ... 7 common frames omitted
```

It appears to be reproducible. This is running as a standalone; I'm not sure how to view the logging for the graphics component. I'll deploy RC4 to staging and test it there. Maybe the graphics logging will provide more information.

Note that, at this point, I'm hoping to deploy on Tuesday; it won't happen today. So no rush. Have a good weekend!

Hank

--- a/build.gradle
+++ b/build.gradle
@@ -609,10 +609,15 @@ project(':wres-writing') {
 
         // Backwards compatibility for JUnit 4 tests
         testImplementation 'junit:junit:4.13.2'
+        testRuntimeOnly 'org.junit.vintage:junit-vintage-engine:5.11.4'
 
         // In-memory implementation of a java.nio.file abstract file system api
         // to test writing of files without actually writing files on the host machine
         testImplementation group: 'com.google.jimfs', name: 'jimfs', version: '1.3.0'
+    }
+
+    test {
+        useJUnitPlatform()
     }
 }
 

--- a/wres-config/src/wres/config/yaml/DeclarationValidator.java
+++ b/wres-config/src/wres/config/yaml/DeclarationValidator.java
@@ -5638,7 +5638,16 @@ public class DeclarationValidator
                                                            SourceInterface.USGS_NWIS,
                                                            SourceInterface.WRDS_AHPS,
                                                            SourceInterface.WRDS_NWM,
-                                                           SourceInterface.WRDS_HEFS );
+                                                           SourceInterface.WRDS_HEFS )
+               && dataset.sources()
+                         .stream()
+                         .anyMatch( s -> Objects.nonNull( s )
+                                         && Objects.nonNull( s.uri() )
+                                         && Objects.nonNull( s.uri()
+                                                              .getScheme() )
+                                         && s.uri()
+                                             .getScheme()
+                                             .startsWith( "http" ) );
     }
 
     /**

--- a/wres-config/test/wres/config/yaml/DeclarationValidatorTest.java
+++ b/wres-config/test/wres/config/yaml/DeclarationValidatorTest.java
@@ -330,12 +330,15 @@ class DeclarationValidatorTest
     {
         Source source = SourceBuilder.builder()
                                      .sourceInterface( SourceInterface.USGS_NWIS )
+                                     .uri( URI.create( "http://foo" ) )
                                      .build();
         Source anotherSource = SourceBuilder.builder()
                                             .sourceInterface( SourceInterface.WRDS_NWM )
+                                            .uri( URI.create( "http://foo" ) )
                                             .build();
         Source yetAnotherSource = SourceBuilder.builder()
                                                .sourceInterface( SourceInterface.WRDS_AHPS )
+                                               .uri( URI.create( "http://foo" ) )
                                                .build();
         Dataset left = DatasetBuilder.builder()
                                      .sources( List.of( source ) )
@@ -1115,6 +1118,7 @@ class DeclarationValidatorTest
     {
         Source source = SourceBuilder.builder()
                                      .sourceInterface( SourceInterface.USGS_NWIS )
+                                     .uri( URI.create( "http://foo" ) )
                                      .build();
 
         Dataset dataset = DatasetBuilder.builder()

--- a/wres-reading/src/wres/reading/nwis/NwisReader.java
+++ b/wres-reading/src/wres/reading/nwis/NwisReader.java
@@ -90,7 +90,7 @@ public class NwisReader implements TimeSeriesReader
     /**
      * @see #of(EvaluationDeclaration, SystemSettings)
      * @param systemSettings the system settings
-     * @return an instance that does not performing any chunking of the time-series data
+     * @return an instance that does not perform any chunking of the time-series data
      * @throws NullPointerException if the systemSettings is null
      */
 

--- a/wres-vis/test/wres/vis/client/GraphicsConsumerFactoryTest.java
+++ b/wres-vis/test/wres/vis/client/GraphicsConsumerFactoryTest.java
@@ -179,11 +179,11 @@ class GraphicsConsumerFactoryTest
         // Clean up by deleting the paths written
         for ( Path next : actualPathsWritten )
         {
-            boolean ignored = next.toFile()
+            boolean ignored = next.toFile() // NOSONAR
                                   .delete();
         }
 
-        boolean ignored = basePath.toFile()
+        boolean ignored = basePath.toFile() // NOSONAR
                                   .delete();
     }
 

--- a/wres-writing/src/wres/writing/csv/statistics/CsvStatisticsWriter.java
+++ b/wres-writing/src/wres/writing/csv/statistics/CsvStatisticsWriter.java
@@ -1251,11 +1251,13 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
             DurationScoreMetricComponent metricComponent = next.getMetric();
 
             // Add the metric name, pretty printed
-            MetricConstants namedMetric = MetricConstants.valueOf( metric.getName().name() );
+            MetricConstants namedMetric = MetricConstants.valueOf( metric.getName()
+                                                                         .name() );
             this.append( joiner, namedMetric.toString(), false );
 
             // Add the metric component name, pretty printed
-            MetricConstants namedMetricComponent = MetricConstants.valueOf( metricComponent.getName().name() );
+            MetricConstants namedMetricComponent = MetricConstants.valueOf( metricComponent.getName()
+                                                                                           .name() );
             this.append( joiner, namedMetricComponent.toString(), false );
 
             // Name qualifier
@@ -1493,51 +1495,33 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
         if ( Objects.nonNull( summaryStatistic )
              && metric.getStatisticName() != MetricName.UNDEFINED )
         {
-            metricName = metric.getStatisticName()
-                               .name()
-                               .replace( "_", " " );
-            metricComponentName = metric.getStatisticComponentName()
-                                        .name()
-                                        .replace( "_", " " );
+            metricName = this.getMetricNamePretty( metric.getStatisticName() );
+            metricComponentName = this.getMetricNamePretty( metric.getStatisticComponentName() );
             units = metricComponent.getUnits();
             minimum = String.valueOf( metric.getStatisticMinimum() );
             maximum = String.valueOf( metric.getStatisticMaximum() );
             optimum = String.valueOf( metric.getStatisticOptimum() );
-            summaryStatisticMetricName = metric.getName()
-                                               .name()
-                                               .replace( "_", " " );
-            summaryStatisticMetricComponentName = metricComponent.getName()
-                                                                 .name()
-                                                                 .replace( "_", " " );
+            summaryStatisticMetricName = this.getMetricNamePretty( metric.getName() );
+            summaryStatisticMetricComponentName = this.getMetricNamePretty( metricComponent.getName() );
         }
         // Diagram that contains summary statistic values, such as a quantile-quantile diagram whose quantiles
         // represent a mean value across geographic features
         else if ( Objects.nonNull( summaryStatistic ) )
         {
-            metricName = metric.getName()
-                               .name()
-                               .replace( "_", " " );
-            metricComponentName = metricComponent.getName()
-                                                 .name()
-                                                 .replace( "_", " " );
+            metricName = this.getMetricNamePretty( metric.getName() );
+            metricComponentName = this.getMetricNamePretty( metricComponent.getName() );
             units = metricComponent.getUnits();
             minimum = String.valueOf( metricComponent.getMinimum() );
             maximum = String.valueOf( metricComponent.getMaximum() );
             optimum = "";
-            summaryStatisticMetricName = summaryStatistic.getStatistic()
-                                                         .name()
-                                                         .replace( "_", " " );
+            summaryStatisticMetricName = this.getStatisticNamePretty( summaryStatistic.getStatistic() );
             summaryStatisticMetricComponentName = "";
         }
         // Diagram is a raw metric (e.g., quantile-quantile diagram for one geographic feature)
         else
         {
-            metricName = metric.getName()
-                               .name()
-                               .replace( "_", " " );
-            metricComponentName = metricComponent.getName()
-                                                 .name()
-                                                 .replace( "_", " " );
+            metricName = this.getMetricNamePretty( metric.getName() );
+            metricComponentName = this.getMetricNamePretty( metricComponent.getName() );
             units = metricComponent.getUnits();
             minimum = String.valueOf( metricComponent.getMinimum() );
             maximum = String.valueOf( metricComponent.getMaximum() );
@@ -1589,6 +1573,8 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
         // variable, all tied together with the same statistics group number for each row.
         for ( Pairs.TimeSeriesOfPairs nextSeries : statistic.getTimeSeriesList() )
         {
+            String metricName = this.getMetricNamePretty( metric.getName() );
+
             // Write the reference times
             for ( ReferenceTime time : nextSeries.getReferenceTimesList() )
             {
@@ -1598,8 +1584,7 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
                                                      .mapToDouble( t -> ( double ) t.getSeconds() )
                                                      .toArray();
 
-                MetricDetails details = new MetricDetails( metric.getName()
-                                                                 .toString(),
+                MetricDetails details = new MetricDetails( metricName,
                                                            time.getReferenceTimeType()
                                                                .toString(),
                                                            SECONDS_SINCE_1970_01_01_T_00_00_00_Z,
@@ -1624,8 +1609,7 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
                                             .mapToDouble( t -> ( double ) t.getSeconds() )
                                             .toArray();
 
-            MetricDetails details = new MetricDetails( metric.getName()
-                                                             .toString(),
+            MetricDetails details = new MetricDetails( metricName,
                                                        "VALID TIME",
                                                        SECONDS_SINCE_1970_01_01_T_00_00_00_Z,
                                                        T_00_00_00_Z,
@@ -1644,8 +1628,7 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
             // Write the variable values, starting with the left variables then the right
             for ( int i = 0; i < leftCount; i++ )
             {
-                MetricDetails nextDetails = new MetricDetails( metric.getName()
-                                                                     .toString(),
+                MetricDetails nextDetails = new MetricDetails( metricName,
                                                                leftNames.get( i ),
                                                                metric.getUnits(),
                                                                NEGATIVE_INFINITY_STRING,
@@ -1671,8 +1654,7 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
             {
                 final int finalI = i;
 
-                MetricDetails nextDetails = new MetricDetails( metric.getName()
-                                                                     .toString(),
+                MetricDetails nextDetails = new MetricDetails( metricName,
                                                                rightNames.get( i ),
                                                                metric.getUnits(),
                                                                NEGATIVE_INFINITY_STRING,
@@ -1742,7 +1724,8 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
 
             // Name qualifier: use an explicit name qualifier first, else the ensemble average type for
             // single-valued metrics. These two things do not overlap.
-            MetricConstants metricName = MetricConstants.valueOf( details.metricName() );
+            MetricConstants metricName = MetricConstants.valueOf( details.metricName()
+                                                                         .replace( " ", "_" ) );
             String qualifier = this.getMetricComponentQualifier( metricName, ensembleAverageType );
 
             this.append( joiner, qualifier, false );
@@ -2086,16 +2069,10 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
         if ( Objects.nonNull( summaryStatistic )
              && metric.getStatisticName() != MetricName.UNDEFINED )
         {
-            metricName = metric.getStatisticName()
-                               .name()
-                               .replace( "_", " " );
-            metricComponentName = metric.getStatisticComponentName()
-                                        .name()
-                                        .replace( "_", " " );
+            metricName = this.getMetricNamePretty( metric.getStatisticName() );
+            metricComponentName = this.getMetricNamePretty( metric.getStatisticComponentName() );
 
-            summaryStatisticMetricName = metric.getName()
-                                               .name()
-                                               .replace( "_", " " );
+            summaryStatisticMetricName = this.getMetricNamePretty( metric.getName() );
             summaryStatisticMetricComponentName = boxplotElement.metricComponentName()
                                                                 .replace( "_", " " );
         }
@@ -2103,22 +2080,16 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
         // represent a mean value across geographic features
         else if ( Objects.nonNull( summaryStatistic ) )
         {
-            metricName = metric.getName()
-                               .name()
-                               .replace( "_", " " );
+            metricName = this.getMetricNamePretty( metric.getName() );
             metricComponentName = boxplotElement.metricComponentName()
                                                 .replace( "_", " " );
-            summaryStatisticMetricName = summaryStatistic.getStatistic()
-                                                         .name()
-                                                         .replace( "_", " " );
+            summaryStatisticMetricName = this.getStatisticNamePretty( summaryStatistic.getStatistic() );
             summaryStatisticMetricComponentName = "";
         }
         // Diagram is a raw metric (e.g., quantile-quantile diagram for one geographic feature)
         else
         {
-            metricName = metric.getName()
-                               .name()
-                               .replace( "_", " " );
+            metricName = this.getMetricNamePretty( metric.getName() );
             metricComponentName = boxplotElement.metricComponentName()
                                                 .replace( "_", " " );
             summaryStatisticMetricName = "";
@@ -2304,9 +2275,33 @@ public class CsvStatisticsWriter implements Function<Statistics, Set<Path>>, Clo
     private boolean hasQuantileSamplingUncertainty( SummaryStatistic summaryStatistic )
     {
         return Objects.nonNull( summaryStatistic )
-               && !summaryStatistic.getDimensionList()
-                                   .contains( SummaryStatistic.StatisticDimension.RESAMPLED )
+               && summaryStatistic.getDimensionList()
+                                  .contains( SummaryStatistic.StatisticDimension.RESAMPLED )
                && summaryStatistic.getStatistic() == SummaryStatistic.StatisticName.QUANTILE;
+    }
+
+    /**
+     * Returns a pretty string representation of the metric name.
+     * @param metricName the metric name
+     * @return a pretty string representation
+     */
+
+    private String getMetricNamePretty( MetricName metricName )
+    {
+        return metricName.name()
+                         .replace( "_", " " );
+    }
+
+    /**
+     * Returns a pretty string representation of the statistic name.
+     * @param statisticName the statistic name
+     * @return a pretty string representation
+     */
+
+    private String getStatisticNamePretty( SummaryStatistic.StatisticName statisticName )
+    {
+        return statisticName.name()
+                            .replace( "_", " " );
     }
 
     /**

--- a/wres-writing/test/wres/writing/client/WritingConsumerFactoryTest.java
+++ b/wres-writing/test/wres/writing/client/WritingConsumerFactoryTest.java
@@ -169,20 +169,19 @@ class WritingConsumerFactoryTest
         }
 
         // Make assertions about the writing written by the single external subscription.
-        Set<Path> expectedPaths = Set.of( basePath.resolve( "DRRC2_DRRC2_DRRC2_HEFS_MEAN_ABSOLUTE_ERROR.png" ),
-                                          basePath.resolve( "DRRC2_DRRC2_DRRC2_HEFS_MEAN_ERROR.png" ),
-                                          basePath.resolve( "DRRC2_DRRC2_DRRC2_HEFS_MEAN_SQUARE_ERROR.png" ) );
+        Set<Path> expectedPaths = Set.of( basePath.resolve( "evaluation.csv.gz" ),
+                                          basePath.resolve( "evaluation.csvt" ) );
 
         Assertions.assertEquals( expectedPaths, actualPathsWritten );
 
         // Clean up by deleting the paths written
         for ( Path next : actualPathsWritten )
         {
-            boolean ignored = next.toFile()
+            boolean ignored = next.toFile() // NOSONAR
                                   .delete();
         }
 
-        boolean ignored = basePath.toFile()
+        boolean ignored = basePath.toFile() // NOSONAR
                                   .delete();
     }
 
@@ -201,7 +200,7 @@ class WritingConsumerFactoryTest
                                                                                     ThresholdOperator.GREATER,
                                                                                     ThresholdOrientation.LEFT ) );
 
-        Geometry geometry = MessageUtilities.getGeometry( "DRRC2", null, null, "POINT ( 23.45, 56.21 )" );
+        Geometry geometry = MessageUtilities.getGeometry( "DRRC2", null, null, "POINT (23.45 56.21)" );
 
         Instant earliestValid = Instant.parse( "2551-03-20T01:00:00Z" );
         Instant latestValid = Instant.parse( "2551-03-20T12:00:00Z" );

--- a/wres-writing/test/wres/writing/csv/statistics/CsvStatisticsWriterTest.java
+++ b/wres-writing/test/wres/writing/csv/statistics/CsvStatisticsWriterTest.java
@@ -522,7 +522,7 @@ class CsvStatisticsWriterTest
                                      + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,"
                                      + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,PT1H,"
                                      + "PT1H,,,,,,-Infinity,,,,,LEFT,GREATER,,,,,,,,,MEAN ABSOLUTE ERROR,"
-                                     + "MAIN,,,0.0,Infinity,0.0,1,QUANTILE,,,RESAMPLED,,0.95,3.0";
+                                     + "MAIN,,,0.0,Infinity,0.0,1,QUANTILE,,,\"RESAMPLED\",,0.95,3.0";
 
             assertEquals( lineOneExpected, actual.get( 1 ) );
 
@@ -530,7 +530,7 @@ class CsvStatisticsWriterTest
                                      + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,"
                                      + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,PT1H,PT1H,"
                                      + ",,,,,-Infinity,,,,,LEFT,GREATER,,,,,,,,,MEAN ERROR,MAIN,,,-Infinity,"
-                                     + "Infinity,0.0,2,QUANTILE,,,RESAMPLED,,0.95,2.0";
+                                     + "Infinity,0.0,2,QUANTILE,,,\"RESAMPLED\",,0.95,2.0";
 
             assertEquals( lineTwoExpected, actual.get( 2 ) );
 
@@ -538,7 +538,7 @@ class CsvStatisticsWriterTest
                                        + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,"
                                        + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,PT1H,PT1H,"
                                        + ",,,,,-Infinity,,,,,LEFT,GREATER,,,,,,,,,MEAN SQUARE ERROR,MAIN,,,0.0,"
-                                       + "Infinity,0.0,3,QUANTILE,,,RESAMPLED,,0.95,1.0";
+                                       + "Infinity,0.0,3,QUANTILE,,,\"RESAMPLED\",,0.95,1.0";
 
             assertEquals( lineThreeExpected, actual.get( 3 ) );
         }
@@ -591,7 +591,7 @@ class CsvStatisticsWriterTest
                                      + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,"
                                      + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,PT1H,"
                                      + "PT1H,,,,,,-Infinity,,,,,LEFT,GREATER,,,,,,,,,MEAN ABSOLUTE ERROR,"
-                                     + "MAIN,,,0.0,Infinity,0.0,1,MAXIMUM,,,FEATURES,,,3.0";
+                                     + "MAIN,,,0.0,Infinity,0.0,1,MAXIMUM,,,\"FEATURES\",,,3.0";
 
             assertEquals( lineOneExpected, actual.get( 1 ) );
 
@@ -599,7 +599,7 @@ class CsvStatisticsWriterTest
                                      + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,"
                                      + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,PT1H,PT1H,"
                                      + ",,,,,-Infinity,,,,,LEFT,GREATER,,,,,,,,,MEAN ERROR,MAIN,,,-Infinity,"
-                                     + "Infinity,0.0,2,MAXIMUM,,,FEATURES,,,2.0";
+                                     + "Infinity,0.0,2,MAXIMUM,,,\"FEATURES\",,,2.0";
 
             assertEquals( lineTwoExpected, actual.get( 2 ) );
 
@@ -607,7 +607,7 @@ class CsvStatisticsWriterTest
                                        + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,"
                                        + "-1000000000-01-01T00:00:00Z,+1000000000-12-31T23:59:59.999999999Z,PT1H,PT1H,"
                                        + ",,,,,-Infinity,,,,,LEFT,GREATER,,,,,,,,,MEAN SQUARE ERROR,MAIN,,,0.0,"
-                                       + "Infinity,0.0,3,MAXIMUM,,,FEATURES,,,1.0";
+                                       + "Infinity,0.0,3,MAXIMUM,,,\"FEATURES\",,,1.0";
 
             assertEquals( lineThreeExpected, actual.get( 3 ) );
         }
@@ -647,7 +647,7 @@ class CsvStatisticsWriterTest
             // Make selected assertions
             String lineOneExpected = "QINE,SQIN,,,1,RIGHT,,\"DRRC2\",,,,\"DRRC2\",,,,\"DRRC2\",,,,2551-03-19T00:00:00Z,"
                                      + "2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,"
-                                     + "MEAN,,,,11.94128,,,0.9,,LEFT,GREATER EQUAL,,,,,,,,,TIME_SERIES_PLOT,T0,,"
+                                     + "MEAN,,,,11.94128,,,0.9,,LEFT,GREATER EQUAL,,,,,,,,,TIME SERIES PLOT,T0,,"
                                      + "SECONDS SINCE 1970-01-01T00:00:00Z,0001-01-01T00:00:00Z,9999-12-31T23:59:59Z,,"
                                      + "1,,,,,,,1.23E7";
 
@@ -656,7 +656,7 @@ class CsvStatisticsWriterTest
             String lineThreeExpected = "QINE,SQIN,,,1,RIGHT,,\"DRRC2\",,,,\"DRRC2\",,,,\"DRRC2\",,,,"
                                        + "2551-03-19T00:00:00Z,2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,"
                                        + "2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,MEAN,,,,11.94128,,,0.9,,LEFT,"
-                                       + "GREATER EQUAL,,,,,,,,,TIME_SERIES_PLOT,VALID TIME,,"
+                                       + "GREATER EQUAL,,,,,,,,,TIME SERIES PLOT,VALID TIME,,"
                                        + "SECONDS SINCE 1970-01-01T00:00:00Z,0001-01-01T00:00:00Z,9999-12-31T23:59:59Z,"
                                        + ",1,,,,,,,1.23036E7";
 
@@ -664,7 +664,7 @@ class CsvStatisticsWriterTest
 
             String lineSixExpected = "QINE,SQIN,,,1,RIGHT,,\"DRRC2\",,,,\"DRRC2\",,,,\"DRRC2\",,,,2551-03-19T00:00:00Z,"
                                      + "2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,"
-                                     + "MEAN,,,,11.94128,,,0.9,,LEFT,GREATER EQUAL,,,,,,,,,TIME_SERIES_PLOT,observed,,"
+                                     + "MEAN,,,,11.94128,,,0.9,,LEFT,GREATER EQUAL,,,,,,,,,TIME SERIES PLOT,observed,,"
                                      + ",-Infinity,Infinity,,2,,,,,,,12.0";
 
             assertEquals( lineSixExpected, actual.get( 6 ) );
@@ -672,40 +672,41 @@ class CsvStatisticsWriterTest
             String lineEightExpected = "QINE,SQIN,,,1,RIGHT,,\"DRRC2\",,,,\"DRRC2\",,,,\"DRRC2\",,,,"
                                        + "2551-03-19T00:00:00Z,2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,"
                                        + "2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,MEAN,,,,11.94128,,,0.9,,LEFT,"
-                                       + "GREATER EQUAL,,,,,,,,,TIME_SERIES_PLOT,predicted,,,-Infinity,Infinity,,2,,,,"
+                                       + "GREATER EQUAL,,,,,,,,,TIME SERIES PLOT,predicted,,,-Infinity,Infinity,,2,,,,"
                                        + ",,,15.7";
 
             assertEquals( lineEightExpected, actual.get( 8 ) );
 
             String lineTenExpected = "QINE,SQIN,,,1,RIGHT,,\"DRRC2\",,,,\"DRRC2\",,,,\"DRRC2\",,,,2551-03-19T00:00:00Z,"
                                      + "2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,"
-                                     + "MEAN,,,,11.94128,,,0.9,,LEFT,GREATER EQUAL,,,,,,,,,TIME_SERIES_PLOT,T0,,"
+                                     + "MEAN,,,,11.94128,,,0.9,,LEFT,GREATER EQUAL,,,,,,,,,TIME SERIES PLOT,T0,,"
                                      + "SECONDS SINCE 1970-01-01T00:00:00Z,0001-01-01T00:00:00Z,9999-12-31T23:59:59Z,,"
                                      + "4,,,,,,,1.23E7";
 
             assertEquals( lineTenExpected, actual.get( 10 ) );
 
             String lineTwelveExpected = "QINE,SQIN,,,1,RIGHT,,\"DRRC2\",,,,\"DRRC2\",,,,\"DRRC2\",,,,"
-                                       + "2551-03-19T00:00:00Z,2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,"
-                                       + "2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,MEAN,,,,11.94128,,,0.9,,LEFT,"
-                                       + "GREATER EQUAL,,,,,,,,,TIME_SERIES_PLOT,VALID TIME,,"
-                                       + "SECONDS SINCE 1970-01-01T00:00:00Z,0001-01-01T00:00:00Z,9999-12-31T23:59:59Z,"
-                                       + ",4,,,,,,,1.23072E7";
+                                        + "2551-03-19T00:00:00Z,2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,"
+                                        + "2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,MEAN,,,,11.94128,,,0.9,,LEFT,"
+                                        + "GREATER EQUAL,,,,,,,,,TIME SERIES PLOT,VALID TIME,,"
+                                        + "SECONDS SINCE 1970-01-01T00:00:00Z,0001-01-01T00:00:00Z,9999-12-31T23:59:59Z,"
+                                        + ",4,,,,,,,1.23072E7";
 
             assertEquals( lineTwelveExpected, actual.get( 12 ) );
 
-            String lineFifteenExpected = "QINE,SQIN,,,1,RIGHT,,\"DRRC2\",,,,\"DRRC2\",,,,\"DRRC2\",,,,2551-03-19T00:00:00Z,"
-                                     + "2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,"
-                                     + "MEAN,,,,11.94128,,,0.9,,LEFT,GREATER EQUAL,,,,,,,,,TIME_SERIES_PLOT,predicted,,"
-                                     + ",-Infinity,Infinity,,3,,,,,,,117.6";
+            String lineFifteenExpected =
+                    "QINE,SQIN,,,1,RIGHT,,\"DRRC2\",,,,\"DRRC2\",,,,\"DRRC2\",,,,2551-03-19T00:00:00Z,"
+                    + "2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,"
+                    + "MEAN,,,,11.94128,,,0.9,,LEFT,GREATER EQUAL,,,,,,,,,TIME SERIES PLOT,predicted,,"
+                    + ",-Infinity,Infinity,,3,,,,,,,117.6";
 
             assertEquals( lineFifteenExpected, actual.get( 15 ) );
 
             String lineSeventeenExpected = "QINE,SQIN,,,1,RIGHT,,\"DRRC2\",,,,\"DRRC2\",,,,\"DRRC2\",,,,"
-                                       + "2551-03-19T00:00:00Z,2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,"
-                                       + "2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,MEAN,,,,11.94128,,,0.9,,LEFT,"
-                                       + "GREATER EQUAL,,,,,,,,,TIME_SERIES_PLOT,predicted,,,-Infinity,Infinity,,4,,,,"
-                                       + ",,,115.7";
+                                           + "2551-03-19T00:00:00Z,2551-03-19T12:00:00Z,2551-03-20T01:00:00Z,"
+                                           + "2551-03-20T12:00:00Z,PT1H,PT7H,PT1H,MEAN,,,,11.94128,,,0.9,,LEFT,"
+                                           + "GREATER EQUAL,,,,,,,,,TIME SERIES PLOT,predicted,,,-Infinity,Infinity,,4,,,,"
+                                           + ",,,115.7";
 
             assertEquals( lineSeventeenExpected, actual.get( 16 ) );
         }

--- a/wres-writing/test/wres/writing/netcdf/NetcdfOutputWriterTest.java
+++ b/wres-writing/test/wres/writing/netcdf/NetcdfOutputWriterTest.java
@@ -256,9 +256,17 @@ class NetcdfOutputWriterTest
                                                               .setSeconds( 10_800 ) )
                               .build();
 
+            TimeWindow timeWindowTwo = timeWindowOne.toBuilder()
+                                                    .setEarliestLeadDuration( Duration.newBuilder()
+                                                                                      .setSeconds( 10_800 ) )
+                                                    .setLatestLeadDuration( Duration.newBuilder()
+                                                                                    .setSeconds( 21_600 ) )
+                                                    .build();
+
             writer.createBlobsForWriting( featureGroups,
                                           metricsAndThresholdsList,
-                                          Set.of( TimeWindowOuter.of( timeWindowOne ) ) );
+                                          Set.of( TimeWindowOuter.of( timeWindowOne ),
+                                                  TimeWindowOuter.of( timeWindowTwo ) ) );
 
             // Create the statistics to write
             DoubleScoreMetric.DoubleScoreMetricComponent
@@ -294,11 +302,7 @@ class NetcdfOutputWriterTest
 
             Pool poolTwo = Pool.newBuilder()
                                .setGeometryGroup( geoGroupTwo )
-                               .setTimeWindow( timeWindowOne.toBuilder()
-                                                            .setEarliestLeadDuration( Duration.newBuilder()
-                                                                                              .setSeconds( 10_800 ) )
-                                                            .setLatestLeadDuration( Duration.newBuilder()
-                                                                                            .setSeconds( 21_600 ) ) )
+                               .setTimeWindow( timeWindowTwo )
                                .setEventThreshold( thresholdTwo )
                                .setDecisionThreshold( thresholdThree )
                                .build();

--- a/wres-writing/testNonsrc/eventbroker.properties
+++ b/wres-writing/testNonsrc/eventbroker.properties
@@ -1,0 +1,18 @@
+# Set the InitialContextFactory class to use
+java.naming.factory.initial = org.apache.qpid.jms.jndi.JmsInitialContextFactory
+
+# Define the required ConnectionFactory instances
+# connectionfactory.<JNDI-lookup-name> = <URI>
+# If the connection port is the reserved TCP port 0, then use the actual broker port discovered on start-up
+connectionFactory.statisticsFactory = amqp://localhost:0?tcpSendBufferSize=1048576;tcpReceiveBufferSize=1048576;protocols=AMQP;useEpoll=true;amqpCredits=1000;amqpLowCredits=300;amqpMinLargeMessageSize=102400;amqpDuplicateDetection=true
+
+# Configure the necessary Queue and/or Topic objects
+# queue.<JNDI-lookup-name> = <queue-name>
+# topic.<JNDI-lookup-name> = <topic-name>
+# queue.statisticsQueue = queue
+
+# One destination under amq.topic for each type of message/subject
+topic.evaluation = BURL:topic://amq.topic/evaluation
+topic.status = BURL:topic://amq.topic/status
+topic.statistics = BURL:topic://amq.topic/statistics
+topic.pairs = BURL:topic://amq.topic/pairs

